### PR TITLE
Weighted flux qoi

### PR DESCRIPTION
**Pull request by roystgnr**
I want to add regression tests before merging, but the code should be in decent shape to examine.

### CODE REVIEW

File: src/qoi/src/weighted_flux_qoi.C [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+    remapped_func->attach_subfunction(raw_func, var_indices);
+
+    _adjoint_bc.reset(remapped_func);
+
+    libMesh::DirichletBoundary adjoint_bc
+      (_bc_ids, var_indices, _adjoint_bc.get());
+
+    // FIXME: this is an ugly hack
+    MultiphysicsSystem & hacked_system =
```
Comment (pbauman) [inline src/qoi/src/weighted_flux_qoi.C]: Feel free to update the CompositeQoI (and, hence, others) to pass a non-const system. There's no good reason for it at this point. I guess my original thinking was we were only doing stuff to the QoIs and not to the system, but obviously that was shortsighted.

Comment (roystgnr) [inline src/qoi/src/weighted_flux_qoi.C]: IMHO passing a non-const system would *also* be an ugly hack; aesthetically you made the right choice.  If you think dropping that const is the less ugly option then I won't argue, though.

**Comment by roystgnr** [comment]
Rebased with merge conflicts fixed.  "Not sure it's working" conflicts are still there.

File: src/qoi/include/grins/qoi_base.h [hunk 2]
```diff
@@ -78,7 +78,9 @@ namespace GRINS
      * Method to allow QoI to cache any system information needed for QoI calculation,
      * for example, solution variable indices.
      */
-    virtual void init( const GetPot& input, const MultiphysicsSystem& system );
+    virtual void init( const GetPot& input,
+                       const MultiphysicsSystem& system,
+                       unsigned int qoi_num );
```
Comment (dmcdougall) [inline src/qoi/include/grins/qoi_base.h]: Does this break backwards compatibility?  `init` is a public method.

Comment (pbauman) [inline src/qoi/include/grins/qoi_base.h]: Yes, but it's effectively internal --- others subclasses would have to be updated if needed.

Comment (roystgnr) [inline src/qoi/include/grins/qoi_base.h]: It's internal, but it's already led to internal bugs in this branch.  Thanks @dmcdougall for bringing it to my attention!

Comment (roystgnr) [inline src/qoi/include/grins/qoi_base.h]: Also?  We totally need to add libmesh_move, libmesh_final, libmesh_override, etc. for all the functions and keywords where C++11 gives you more efficiency or better error checking but where C++98 still compiles to correct code with a no-op function or an empty macro.

Comment (pbauman) [inline src/qoi/include/grins/qoi_base.h]: I agree, but I'm not up-to-speed. How would that play in this particular case (just passing in references, not copying data)?

Comment (roystgnr) [inline src/qoi/include/grins/qoi_base.h]: In this case I'm not talking about move-vs-copy, I'm talking about override.  The ParsedBoundary QoI was a separate development thread, so it didn't get the API change that is in this thread, so although git thought rebasing off master went without a hitch, it actually created a runtime error.  override would have caught the mistake at compile-time.

Or libmesh_override, naturally.  See https://github.com/libMesh/libmesh/issues/655 if you want to chime in on that.

**Comment by roystgnr** [comment]
This is giving me good results finally; thanks to @vikramvgarg for finding the first libMesh bug and helping find the second in https://github.com/libMesh/libmesh/pull/654

We should be ready to merge; this will compile with pre-654 libMesh, it just won't give correct results except in a few trivial cases.

**Comment by vikramvgarg** [comment]
Cool ! I'll try it out on the airfoil example.

On Mon, Sep 21, 2015 at 9:33 PM, roystgnr <notifications@github.com> wrote:

> This is giving me good results finally; thanks to @vikramvgarg
> <https://github.com/vikramvgarg> for finding the first libMesh bug and
> helping find the second in libMesh/libmesh#654
> <https://github.com/libMesh/libmesh/pull/654>
>
> We should be ready to merge; this will compile with pre-654 libMesh, it
> just won't give correct results except in a few trivial cases.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grinsfem/grins/pull/275#issuecomment-142162339>.
>



-- 
Vikram Garg
Postdoctoral Associate
Predictive Engineering and Computational Science (PECOS)
The University of Texas at Austin
http://web.mit.edu/vikramvg/www/

http://vikramvgarg.wordpress.com/
http://www.runforindia.org/runners/vikramg

**Comment by pbauman** [comment]
OK, rockin'. Once @vikramvgarg reports on his airfoil example, let's merge.

**Comment by vikramvgarg** [comment]
It seems that the adjoint solution is not being correctly computed for the
airfoil. The adjoint solution looks very different from the my earlier
brute force implementation earlier. And the drag computed has a value that
is quite a bit off from direct evaluation:

Weighted Flux: 2.2005847397264782e-03
Direct: 4.0252425243265857e-03

On Tue, Sep 22, 2015 at 7:46 AM, Paul T. Bauman <notifications@github.com>
wrote:

> OK, rockin'. Once @vikramvgarg <https://github.com/vikramvgarg> reports
> on his airfoil example, let's merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grinsfem/grins/pull/275#issuecomment-142277671>.
>



-- 
Vikram Garg
Postdoctoral Associate
Predictive Engineering and Computational Science (PECOS)
The University of Texas at Austin
http://web.mit.edu/vikramvg/www/

http://vikramvgarg.wordpress.com/
http://www.runforindia.org/runners/vikramg

**Comment by vikramvgarg** [comment]
The problem was in my primal solution, specifically some unnecessary
pressure pinning.

The weighted flux now (-4.0696116018017455e-03) matches the benchmark
calculation on a NASA website to a dot. Should be good to merge.

On Tue, Sep 22, 2015 at 6:19 PM, Vikram Garg <vikram.v.garg@gmail.com>
wrote:

> It seems that the adjoint solution is not being correctly computed for the
> airfoil. The adjoint solution looks very different from the my earlier
> brute force implementation earlier. And the drag computed has a value that
> is quite a bit off from direct evaluation:
>
> Weighted Flux: 2.2005847397264782e-03
> Direct: 4.0252425243265857e-03
>
> On Tue, Sep 22, 2015 at 7:46 AM, Paul T. Bauman <notifications@github.com>
> wrote:
>
>> OK, rockin'. Once @vikramvgarg <https://github.com/vikramvgarg> reports
>> on his airfoil example, let's merge.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/grinsfem/grins/pull/275#issuecomment-142277671>.
>>
>
>
>
> --
> Vikram Garg
> Postdoctoral Associate
> Predictive Engineering and Computational Science (PECOS)
> The University of Texas at Austin
> http://web.mit.edu/vikramvg/www/
>
> http://vikramvgarg.wordpress.com/
> http://www.runforindia.org/runners/vikramg
>



-- 
Vikram Garg
Postdoctoral Associate
Predictive Engineering and Computational Science (PECOS)
The University of Texas at Austin
http://web.mit.edu/vikramvg/www/

http://vikramvgarg.wordpress.com/
http://www.runforindia.org/runners/vikramg

**Comment by roystgnr** [comment]
5 months I let that sit; about time we got it fixed!  Thanks again @vikramvgarg

**Pull request merged**